src/coordsim/traffic_predictor/lstm_predictor.py

```diff
@@ -0,0 +1,301 @@
+import pandas as pd
+from sklearn.preprocessing import MinMaxScaler
+from keras.models import Sequential, load_model
+from keras.layers import Dense
+from keras.layers import LSTM
+from coordsim.reader import reader
+import numpy as np
+import random
+from pickle import load, dump
+import os
+import argparse
+
+
+class LSTM_Predictor:
+    """
+    Traffic prediction module based on LSTM neural networks
+    NOTICE: Only works with Poisson arrival and trace based scenarios with single ingress node
+    Based on work done here:
+    https://machinelearningmastery.com/time-series-forecasting-long-short-term-memory-network-python/
+    """
+    def __init__(self, trace, params, training_repeats=10, nb_epochs=10,
+                 weights_dir=False, poisson_data=False):
+        """
+        Initiate the class
+        PARAMETERS:
+        ----------
+        trace: list: The traffic trace from the simulator. Will be used to generated training data.
+        params: SimulatorParams: the simulator parameters class
+        training_repeats: int: Number of repetitions for the training process.
+        nb_epochs: int: the number of epochs to train the neural network in each repetition.
+        run_duration: the simulator interface's run duration.
+        weights_dir: path to weights dir. Set to false when training
+        poisson: set to True to generate training data randomly based on Poisson process, not the arrival mean
+        """
+        # Store arguments
+        self.trace = trace
+        self.params = params
+        self.training_repeats = training_repeats
+        self.nb_epochs = nb_epochs
+        self.run_duration = self.params.run_duration
+        self.weights_dir = weights_dir
+        if weights_dir:
+            self.weights_dir = os.path.join(os.getcwd(), weights_dir)
+        self.poisson_data = poisson_data
+
+        self.requested_traffic = []  # empty array that will hold training data for the LSTM NN
+        self.training_data = None  # Placeholder for pandas DataFrame to hold training data
+        self.predictions = []
+        self.last_prediction_index = 0  # Index to keep track of position in testing data
+
+        self.gen_training_data()
+        self.prepare_model()
+
+        # ax.plot(self.actual_requested_traffic, label="Actual traffic - seed")
+        # ax.plot(self.requested_traffic, label="Actual traffic - mean")
+        # ax.plot(self.predictions, label="predictions")
+        # fig.tightlayout()
+
+    def gen_training_data(self):
+        self.reset_flow_lists()
+        for i in range(len(self.trace)):
+            cur_time = float(self.trace[i]['time'])
+            inter_arr_mean = float(self.trace[i]['inter_arrival_mean'])
+            if i < len(self.trace) - 1:
+                # Check for multiple runs with the same mean
+                while cur_time < float(self.trace[i+1]['time']):
```

Comment:
The comment seems wrong. The while loop checks that the current line being read has time strictly less than the next one in the trace file. What happens if the condition fails? It just skips that line and moves on to the next one? Maybe log the error/warning?

Reply:
No. It just checks for periods not covered in the trace. E.g. trace jumps from time 500 to time 1000, we still need the run data in between.